README.md
# RADish

```
#      ____  ___    ____  _      __                                                      #
#     / __ \/   |  / __ \(_)____/ /_                                                     #
#    / /_/ / /| | / / / / / ___/ __ \                                                    #
#   / _, _/ ___ |/ /_/ / (__  ) / / /                                                    #
#  /_/ |_/_/  |_/_____/_/____/_/ /_/                                                     #
```

Scripts aiding file processing and analysis of RADseq and GBS data

## LICENSE

All code within the ```RADish``` v0.1.0 repository is available "AS IS" under a generous 3-Clause BSD license. See the [LICENSE](LICENSE) file for more information.

## CITATION

If you use scripts from this repository as part of your published research, I require that you cite the repository as follows (also see DOI information below): 
  
- Bagley, J.C. 2018. RADish v0.1.0. GitHub repository, Available at: http://github.com/justincbagley/RADish.

Alternatively, please provide the following link to this software repository in your manuscript:

- https://github.com/justincbagley/RADish


## USAGE

### fastq_ReadLengthChecker

Example usage code and output to screen during a ```fastq_ReadLengthChecker.sh``` run, in which
reads were discovered to range from 78 bp to 140 bp in length:

```
$ ./fastq_readLengthChecker.sh .

##########################################################################################
#                     fastq_ReadLengthChecker v0.1.0, February 2018                      #
##########################################################################################

INFO      | Wed Feb 21 17:58:12 EST 2018 | Found 1123 .fastq files in current working directory. 
INFO      | Wed Feb 21 17:58:12 EST 2018 | Shortest read length: 78 bp 
INFO      | Wed Feb 21 17:58:12 EST 2018 | Longest read length: 140 bp 
INFO      | Wed Feb 21 17:58:12 EST 2018 | Results output to 'fastq_lengths_summary.txt' in current working directory.
INFO      | Wed Feb 21 17:58:12 EST 2018 | 
INFO      | Wed Feb 21 17:58:12 EST 2018 | ...Cleaning up workspace... 
INFO      | Wed Feb 21 17:58:12 EST 2018 | 
INFO      | Wed Feb 21 17:58:12 EST 2018 | Done checking fastq read lengths. 
INFO      | Wed Feb 21 17:58:12 EST 2018 | Bye.
```

### fastqTrimmer

**Usage**

```
$ fastqTrimmer -h

Usage: $(basename "$0") [Help: -h help] [Options: -i l] workingDir 
 ## Help:
  -h   help text (also: -help)

 ## Options:
  -i   inputFile (def: NULL) file name, in case of single input file
  -b   startingBase (def: 1) nucleotide position to start keeping bases from
       (=also starting position for counting up to trim point)
  -l   trimLength (def: $MY_TRIM_LENGTH) desired final length of reads
  -o   output (def: NULL; e.g. 'output') basename for output .fastq file name
  
 OVERVIEW
 Trims one or more fastq files from their original length to the length (trimLength) 
 desired by the user, by trimming off bases from the right (3') end of sequence reads
 in the fastq file(s). Useful for trimming reads from multiple lanes or assemblies so 
 that they all have the same length prior to calling SNPs or merging assemblies. For 
 example, the author has used this script to prepare fastqs from different lanes of 
 Illumina sequencing on different or mixed sets of samples prior to de novo or 
 reference-based assembly and SNP calling in pyRAD (Eaton 2014) or ipyrad (Eaton and 
 Overcast 2017).
 
 Several options are available. If an input file is specified using the -i flag, then only 
 that file will be trimmed, and if no output basename (-o) is given then the trimmed file 
 will replace the original file. If no input file or output names are given, then the 
 script will trim all .fastq files in the current working directory (final workingDir
 argument).

 CITATION
 Bagley, J.C. 2018. RADish v0.1.0. GitHub repository, Available at: 
	<http://github.com/justincbagley/RADish>.

 REFERENCES
 Eaton DA (2014) PyRAD: assembly of de novo RADseq loci for phylogenetic analyses. 
	Bioinformatics, 30, 1844-1849.
 Eaton DAR, Overcast I (2017) ipyrad: interactive assembly and analysis of RADseq data sets. 
	Available at: <http://ipyrad.readthedocs.io/>.
```

Example usage code and output to screen during a ```fastqTrimmer.sh``` run, in which all fastq
files within the working directory were trimmed down to the first 1-91 bases:


```
$ ./fastqTrimmer.sh -l 91 .

##########################################################################################
#                           fastqTrimmer v0.1.0, February 2018                           #
##########################################################################################

INFO      | Thu Feb 22 09:32:59 MST 2018 | STEP #1: SETUP. 
INFO      | Thu Feb 22 09:32:59 MST 2018 |          Setting working directory to: . 
INFO      | Thu Feb 22 09:32:59 MST 2018 | STEP #2: TRIMMING READS. 
INFO      | Thu Feb 22 09:32:59 MST 2018 |          Trimming reads in all .fastq files in current directory. Trimmed files will be moved to 
INFO      | Thu Feb 22 09:32:59 MST 2018 |          ./trimmed_fastq/, and original fastq files will be moved to ./orig_fastq/.  
./APA1_1.R1.fastq
./APA1_2.R1.fastq
./APA1_3.R1.fastq
./APA1_4.R1.fastq
./APA1_5.R1.fastq
./APA1_6.R1.fastq
./APA1_7.R1.fastq
./APA1_8.R1.fastq
./CAP1H_10.R1.fastq
./CAP1H_1.R1.fastq
./CAP1H_2.R1.fastq
./CAP1H_3.R1.fastq
./CAP1H_4.R1.fastq
./CAP1H_5.R1.fastq
./CAP1H_6.R1.fastq
./CAP1H_7.R1.fastq
./CAP1H_8.R1.fastq
./CAP1H_9.R1.fastq
./CAP1L_10.R1.fastq
./CAP1L_1.R1.fastq
.
.
.
```


February 22, 2018
Justin C. Bagley, Richmond, VA, USA


changeLog.md
# RADish Change Log

## In preparation for RADish v0.1.0 (pre-release version zero)

- **March 2020:** Added new `fastp_Runner.sh` that automates running FASTQ quality control assessment and trimming using the software program `fastp` available [here](https://github.com/OpenGene/fastp), which in my opinion is an improvement over the industry standard software FASTQC (`fastp` is faster and more informative).
- **November 2018:** Added new 'extract_VCF_sampleNames.sh' script that extracts all sample names from a VCF file, and also created this RADish Change Log file for keeping track of future updates before the first pre-release.
- **May 2018:** Updated the README and 'VCF2minor012.sh', 'fastqTrimmer.sh', 'fastq_readLengthChecker.sh', 'removeFixedSNPs.sh', and 'minor012-bayenv2.sh' scripts
- **2017–early 2018:** Created and added initial scripts and README for repo while analyzing ddRAD-seq data as a postdoc in Andrew Eckert's lab at Virginia Commonwealth University (VCU). Initial file list:
  * LICENSE
  * README.md
  * README.pdf
  * VCF2minor012.sh
  * bayenv2MatrixTracer.sh
  * bayenv2SNPSFILESplitter.sh
  * fastqTrimmer.sh
  * fastq_readLengthChecker.sh
  * gzipVCFs.sh
  * minor012-bayenv2.sh
  * removeFixedSNPs.sh
  * toyBayenv2Data.sh
 


fastqTrimmer.sh
#!/bin/sh

##########################################################################################
#      ____  ___    ____  _      __                                                      #
#     / __ \/   |  / __ \(_)____/ /_                                                     #
#    / /_/ / /| | / / / / / ___/ __ \                                                    #
#   / _, _/ ___ |/ /_/ / (__  ) / / /                                                    #
#  /_/ |_/_/  |_/_____/_/____/_/ /_/                                                     #
#                           fastqTrimmer v0.1.0, February 2018                           #
#  SHELL SCRIPT FOR TRIMMING READS IN FASTQ FILES TO USER-SPECIFIED LENGTHS AND ALSO     #
#  STARTING POSITION                                                                     #
#  Copyright ©2018 Justinc C. Bagley. For further information, see README and license    #
#  available in the RADish repository (https://github.com/justincbagley/RADish/). Last   #
#  update: February 22, 2018. For questions, please email jcbagley@vcu.edu.              #
##########################################################################################

############ SCRIPT OPTIONS
## OPTION DEFAULTS ##
MY_FILENAME=NULL
MY_STARTING_BASE=1
MY_TRIM_LENGTH=100
MY_OUTPUT=NULL

############ CREATE USAGE & HELP TEXTS
Usage="Usage: $(basename "$0") [Help: -h help] [Options: -i l] workingDir 
 ## Help:
  -h   help text (also: -help)

 ## Options:
  -i   inputFile (def: NULL) file name, in case of single input file
  -b   startingBase (def: 1) nucleotide position to start keeping bases from
       (=also starting position for counting up to trim point)
  -l   trimLength (def: $MY_TRIM_LENGTH) desired final length of reads
  -o   output (def: NULL; e.g. 'output') basename for output .fastq file name
  
 OVERVIEW
 Trims one or more fastq files from their original length to the length (trimLength) 
 desired by the user, by trimming off bases from the right (3') end of sequence reads
 in the fastq file(s). Useful for trimming reads from multiple lanes or assemblies so 
 that they all have the same length prior to calling SNPs or merging assemblies. For 
 example, the author has used this script to prepare fastqs from different lanes of 
 Illumina sequencing on different or mixed sets of samples prior to de novo or 
 reference-based assembly and SNP calling in pyRAD (Eaton 2014) or ipyrad (Eaton and 
 Overcast 2017).
 
 Several options are available. If an input file is specified using the -i flag, then only 
 that file will be trimmed, and if no output basename (-o) is given then the trimmed file 
 will replace the original file. If no input file or output names are given, then the 
 script will trim all .fastq files in the current working directory (final workingDir
 argument).

 CITATION
 Bagley, J.C. 2018. RADish v0.1.0. GitHub repository, Available at: 
	<http://github.com/justincbagley/RADish>.

 REFERENCES
 Eaton DA (2014) PyRAD: assembly of de novo RADseq loci for phylogenetic analyses. 
	Bioinformatics, 30, 1844-1849.
 Eaton DAR, Overcast I (2017) ipyrad: interactive assembly and analysis of RADseq data sets. 
	Available at: <http://ipyrad.readthedocs.io/>.
"

if [[ "$1" == "-h" ]] || [[ "$1" == "-help" ]]; then
	echo "$Usage"
	exit
fi

############ PARSE THE OPTIONS
while getopts 'i:b:l:o:' opt ; do
  case $opt in

## trimmer options:
    i) MY_FILENAME=$OPTARG ;;
    b) MY_STARTING_BASE=$OPTARG ;;
    l) MY_TRIM_LENGTH=$OPTARG ;;
    o) MY_OUTPUT=$OPTARG ;;

## Missing and illegal options:
    :) printf "Missing argument for -%s\n" "$OPTARG" >&2
       echo "$Usage" >&2
       exit 1 ;;
   \?) printf "Illegal option: -%s\n" "$OPTARG" >&2
       echo "$Usage" >&2
       exit 1 ;;
  esac
done

############ SKIP OVER THE PROCESSED OPTIONS
shift $((OPTIND-1)) 
# Check for mandatory positional parameters
if [ $# -lt 1 ]; then
echo "$Usage"
  exit 1
fi
USER_SPEC_PATH="$1"

echo "
##########################################################################################
#                           fastqTrimmer v0.1.0, February 2018                           #
##########################################################################################
"

######################################## START ###########################################
echo "INFO      | $(date) | STEP #1: SETUP. "
###### Set new path/dir environmental variable to user specified path, then create useful
##--shell functions and variables:
if [[ "$USER_SPEC_PATH" = "$(echo $(pwd))" ]]; then
	MY_PATH=`pwd -P`
	echo "INFO      | $(date) |          Setting working directory to: $MY_PATH "
elif [[ "$USER_SPEC_PATH" != "$(echo $(pwd))" ]]; then
	MY_PATH=$USER_SPEC_PATH
	echo "INFO      | $(date) |          Setting working directory to: $MY_PATH "	
else
	echo "WARNING!  | $(date) |          Null working directory path. Quitting... "
	exit 1
fi
	CR=$(printf '\r'); 
	calc () { 
		bc -l <<< "$@" 
}

##--Move into user specified path (solves problem of executing analysis for non-local
##--working dir):
	cd $MY_PATH


echo "INFO      | $(date) | STEP #2: TRIMMING READS. "
## CASE 1 and 2 - with input name only, or with both input and output specified, respectively.
## Do this with a big if-elif-fi statement, while reading user input to check and doing a 
## conditional subroutine.

## PSEUDO CODE:
##
## if only i is specified; then
##   warning will trim file in place
##   cut -c 1-trimLength ./inputFile.fastq > ./inputFile.fastq.tmp
##   rm ./inputFile.fastq
##   mv ./inputFile.fastq.tmp ./inputFile.fastq
## elif i and o are specified; then
##   cut -c 1-trimLength ./inputFile.fastq > ./outputFile.fastq
## fi

if [[ "$MY_FILENAME" != "NULL" ]] && [[ "$MY_OUTPUT" = "NULL" ]]; then
	echo "WARNING!  | $(date) |          fastqTrimmer will trim the input file in place and replace the old file with the trimmed file. "
	echo ""
	read -p "Do you want to proceed? (0=no, 1=yes) [ def: 0 ] : " TRIM_REPLACE_SWITCH 
	echo ""
#
	if [[ "$TRIM_REPLACE_SWITCH" = 0 ]]; then
		echo "WARNING!  | $(date) |                          ILLEGAL OPTION. Quitting... "
		exit 1
	elif [[ "$TRIM_REPLACE_SWITCH" = 1 ]]; then
		echo "INFO      | $(date) |          Trimming reads in the following file: '$MY_FILENAME'. "
		TRUE_TRIM_POSITION="$(calc $MY_TRIM_LENGTH + $MY_STARTING_BASE - 1)"
		MYBASENAME="$(basename $MY_FILENAME '.fastq')"
		cut -c "$MY_STARTING_BASE"-"$TRUE_TRIM_POSITION" "$MY_FILENAME" > ./"$MYBASENAME".fastq.tmp
		rm "$MY_FILENAME"
		mv ./"$MYBASENAME".fastq.tmp ./"$MYBASENAME".fastq	
	fi
#
elif [[ "$MY_FILENAME" != "NULL" ]] && [[ "$MY_OUTPUT" != "NULL" ]]; then
		echo "INFO      | $(date) |          Trimming reads in file '$MY_FILENAME', and renaming with basename '$MY_OUTPUT'. "
		TRUE_TRIM_POSITION="$(calc $MY_TRIM_LENGTH + $MY_STARTING_BASE - 1)"
		cut -c "$MY_STARTING_BASE"-"$TRUE_TRIM_POSITION" "$MY_FILENAME" > ./"$MY_OUTPUT".fastq
fi



## CASE 3 - perhaps the normal case. Input file and output are left at their NULL defaults
## (not specified) and so we assume a working dir with multiple .fastq files and we loop through
## the fastqs, trim them according to user specifications, and then output '"$MYBASENAME"_trim.fastq'
## style output file names.

## PSEUDO CODE:
## if i = NULL; then
##   (
##     for i in ./*.fastq; do
##       MYBASENAME="$()" 
##       cut -c 1-trimLength ./Strauss_GBS.fastq > ./Strauss_GBS_trim.fastq
##     done
##   )
## fi

if [[ "$MY_FILENAME" = "NULL" ]] && [[ "$MY_OUTPUT" = "NULL" ]]; then
echo "INFO      | $(date) |          Trimming reads in all .fastq files in current directory. Trimmed files will be moved to "
echo "INFO      | $(date) |          ./trimmed_fastq/, and original fastq files will be moved to ./orig_fastq/.  "
(
	for i in ./*.fastq; do
		echo "$i"
		TRUE_TRIM_POSITION="$(calc $MY_TRIM_LENGTH + $MY_STARTING_BASE - 1)"
		MYBASENAME="$(basename $i '.fastq')"
		cut -c "$MY_STARTING_BASE"-"$TRUE_TRIM_POSITION" "$i" > ./"$MYBASENAME"_trim.fastq
	done
)
	mkdir orig_fastq trimmed_fastq
	mv ./*_trim.fastq ./trimmed_fastq/
	mv ./*.fastq ./orig_fastq/
fi


exit 0

